data_model/derive/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+        panic!("Only enums are supported")
+    };
+
+    let mut filter_variants_id_fields = Vec::new();
+    let mut filter_variants_idents_event_fields = Vec::new();
+
+    let filter_variants_event_fields = event_variants.iter().filter_map(|variant| {
+        let event_filter_variant_name = format_ident!("By{}", variant.ident);
+        if let syn::Fields::Unnamed(ref unnamed) = variant.fields {
+            let variant_type = &unnamed
+                .unnamed
+                .first()
+                .expect("Should have at least one type")
+                .ty;
+            if let syn::Type::Path(path) = variant_type {
+                let var_ty_ident = &path.path.segments[0].ident;
+                // Only `...Event` containing fields are wrapped with `FilterOpt`
+                if var_ty_ident.to_string().contains("Event") {
+                    // Also store just the variant ident for later codegen
+                    filter_variants_idents_event_fields.push(event_filter_variant_name.clone());
+                    let inner_filter_ident = format_ident!(
+                        "{}Filter",
+                        var_ty_ident.to_string().strip_suffix("Event").expect("Checked in outer if"),
+                    );
+                    Some(quote! {
+                        #event_filter_variant_name (crate::prelude::FilterOpt<#inner_filter_ident>)
+                    })
+                } else {
+                    // Default case, make a unit variant with just the ident
+                    filter_variants_id_fields.push(event_filter_variant_name);
+                    None
+                }
+            }
+            else { None }
+        }
+        else { None }
+    }).collect::<Vec<_>>();
+
+    let filter_impl_variant_pairs_event_fields = filter_variants_idents_event_fields
+        .iter()
+        .zip(event_variants.iter())
+        .map(|(filter_variant_ident, event_var)| {
+            let event_var_ident = format_ident!("{}", &event_var.ident);
+            quote! {
+                (Self::#filter_variant_ident(filter_opt), crate::prelude::#event_ident::#event_var_ident(event)) => {
+                    filter_opt.matches(event)
+                }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let filter_impl_variant_pairs_id_fields = filter_variants_id_fields
+        .iter()
+        .zip(
+            event_variants
+                .iter()
+                .skip(filter_impl_variant_pairs_event_fields.len()),
+        )
+        .map(|(filter_var, event_var)| {
+            let event_var_ident = format_ident!("{}", &event_var.ident);
+            quote! {
+                (Self::#filter_var, crate::prelude::#event_ident::#event_var_ident (_))
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let filter_doc = format!("Filter for {} entity", filter_ident);
+    let new_doc = format!("Construct new {}", filter_ident);
+
+    quote! {
+        #[derive(
+            Clone,
```

Comment:
I'm in favour of enforcing invariants pre-emptively. Implementing as many traits as possible is also considered a good practice.